pkg/runtime.v2/framework/plugins/jobset/jobset.go

```diff
@@ -0,0 +1,121 @@
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	apiruntime "k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/utils/ptr"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	ctrlutil "sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	jobsetv1alpha2 "sigs.k8s.io/jobset/api/jobset/v1alpha2"
+
+	kubeflowv2 "github.com/kubeflow/training-operator/pkg/apis/kubeflow.org/v2alpha1"
+	runtime "github.com/kubeflow/training-operator/pkg/runtime.v2"
+	"github.com/kubeflow/training-operator/pkg/runtime.v2/framework"
+)
+
+type JobSet struct {
+	client     client.Client
+	restMapper meta.RESTMapper
+	scheme     *apiruntime.Scheme
+}
+
+var _ framework.WatchExtensionPlugin = (*JobSet)(nil)
+var _ framework.ComponentBuilderPlugin = (*JobSet)(nil)
+
+const Name = "JobSet"
+
+func New(_ context.Context, c client.Client, _ client.FieldIndexer) (framework.Plugin, error) {
+	return &JobSet{
+		client:     c,
+		restMapper: c.RESTMapper(),
+		scheme:     c.Scheme(),
+	}, nil
+}
+
+func (j *JobSet) Name() string {
+	return Name
+}
+
+func (j *JobSet) Build(ctx context.Context, info *runtime.Info, trainJob *kubeflowv2.TrainJob) (client.Object, error) {
+	if info == nil || info.Obj == nil || trainJob == nil {
+		return nil, fmt.Errorf("runtime info or object is missing")
+	}
+	raw, ok := info.Obj.(*jobsetv1alpha2.JobSet)
+	if !ok {
+		return nil, nil
+	}
+	jobSetBuilder := NewBuilder(client.ObjectKeyFromObject(trainJob), kubeflowv2.JobSetTemplateSpec{
+		ObjectMeta: metav1.ObjectMeta{
+			Labels:      info.Labels,
+			Annotations: info.Annotations,
+		},
+		Spec: raw.Spec,
+	})
+	jobSet := jobSetBuilder.
+		ContainerImage(trainJob.Spec.Trainer.Image).
+		JobCompletionMode(batchv1.IndexedCompletion).
+		PodLabels(info.PodLabels).
+		Build()
+	if err := ctrlutil.SetControllerReference(trainJob, jobSet, j.scheme); err != nil {
+		return nil, err
+	}
+	oldJobSet := &jobsetv1alpha2.JobSet{}
+	if err := j.client.Get(ctx, client.ObjectKeyFromObject(jobSet), oldJobSet); err != nil {
+		if !apierrors.IsNotFound(err) {
+			return nil, err
+		}
+		oldJobSet = nil
+	}
+	if err := info.Update(jobSet); err != nil {
+		return nil, err
+	}
+	if needsCreateOrUpdate(oldJobSet, jobSet, ptr.Deref(trainJob.Spec.Suspend, false)) {
+		return jobSet, nil
+	}
+	return nil, nil
+}
+
+func needsCreateOrUpdate(old, new *jobsetv1alpha2.JobSet, suspended bool) bool {
+	return old == nil ||
+		suspended && (!equality.Semantic.DeepEqual(old.Spec, new.Spec) || !maps.Equal(old.Labels, new.Labels) || !maps.Equal(old.Annotations, new.Annotations))
+}
+
+func (j *JobSet) ReconcilerBuilders() []runtime.ReconcilerBuilder {
+	if _, err := j.restMapper.RESTMapping(
+		schema.GroupKind{Group: jobsetv1alpha2.GroupVersion.Group, Kind: "JobSet"},
+		jobsetv1alpha2.SchemeGroupVersion.Version,
+	); err != nil {
+		return nil
```

Comment:
This is checking if the JobSet CRD is deployed to the cluster.

Reply:
When we return errors here, the operator will crash during the startup phase. 

So, I want to operate this behavior based on the configuration since the command arg is not clear for the users as today's training-operator v1 (we often questions for arg from users).

Here we keep just returning nil, and then can we create issues to implement Configuration?

Reply:
>Here we keep just returning nil, and then can we create issues to implement Configuration?

That sounds good to me for PodGroup check, however JobSet CRD must be installed for all use-cases, right ?
Thus, maybe if JobSet is not installed we just return an error to the user.

Reply:
> Should we remove this check once we have a use-case when we have runtime that is not using JobSet ?

But, now, they can implement the custom plugin directly here. Instead of removal of this checking, how about the warning logs, and not return nil like this:


```go
	if _, err := j.restMapper.RESTMapping(
		schema.GroupKind{Group: jobsetv1alpha2.GroupVersion.Group, Kind: "JobSet"},
		jobsetv1alpha2.SchemeGroupVersion.Version,
	); err != nil {
		~LOG~
	}
	return []runtime.ReconcilerBuilder{
		func(b *builder.Builder, c client.Client) *builder.Builder {
			return b.Owns(&jobsetv1alpha2.JobSet{})
	
```

Reply:
In the future, we will proper error handling here once we implement Config API to opt-in / opt-out plugin.

Reply:
Yes, maybe adding the Warning message is ok for now.
Let's add the TODO as well.

Reply:
SGTM

Reply:
Done.